projects/prison-custody-status-to-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/messaging/Handler.kt

```diff
@@ -5,52 +5,80 @@ import uk.gov.justice.digital.hmpps.converter.NotificationConverter
+                            mapOf("nomsId" to nomsId)
+                        )
+                    }
+                    prisonerMovement = booking.prisonerMovement(hmppsEvent.occurredAt)
+                    recordMovement(prisonerMovement)
+                }
+
                 "prison-offender-events.prisoner.released" -> {
-                    val outcome = releaseService.release(
-                        hmppsEvent.additionalInformation.nomsNumber(),
-                        hmppsEvent.additionalInformation.prisonId(),
-                        hmppsEvent.additionalInformation.reason(),
-                        hmppsEvent.additionalInformation.movementReason(),
-                        hmppsEvent.occurredAt
-                    )
+                    prisonerMovement = hmppsEvent.asReleased()
+                    val outcome = releaseService.release(prisonerMovement)
                     telemetryService.trackEvent(outcome.name, hmppsEvent.telemetryProperties())
                 }
 
                 "prison-offender-events.prisoner.received" -> {
-                    val outcome = recallService.recall(
-                        hmppsEvent.additionalInformation.nomsNumber(),
-                        hmppsEvent.additionalInformation.prisonId()!!,
-                        hmppsEvent.additionalInformation.reason(),
-                        hmppsEvent.additionalInformation.movementReason(),
-                        hmppsEvent.occurredAt
-                    )
+                    prisonerMovement = hmppsEvent.asReceived()
+                    val outcome = recallService.recall(prisonerMovement)
                     telemetryService.trackEvent(outcome.toString(), hmppsEvent.telemetryProperties())
                 }
 
-                else -> throw IllegalArgumentException("Unknown event type ${notification.eventType}")
+                else -> {
+                    throw IllegalArgumentException("Unknown event type ${hmppsEvent.eventType}")
+                }
             }
         } catch (e: IgnorableMessageException) {
-            telemetryService.trackEvent(e.message, hmppsEvent.telemetryProperties() + e.additionalProperties)
-            return
+            telemetryService.trackEvent(
+                e.message,
+                prisonerMovement.telemetryProperties() + e.additionalProperties
+            )
+        }
+    }
+
+    private fun recordMovement(prisonerMovement: PrisonerMovement) {
+        when (prisonerMovement) {
+            is PrisonerMovement.Received -> {
+                val outcome = recallService.recall(prisonerMovement)
+                telemetryService.trackEvent(outcome.name, prisonerMovement.telemetryProperties())
+            }
+
+            is PrisonerMovement.Released -> {
+                telemetryService.trackEvent("IdentifierAddedForReleasedPrisoner", prisonerMovement.telemetryProperties())
+                throw IllegalStateException("NomsNumberAddedAfterRelease")
+            }
```

Comment:
Shouldn't this only be thrown if they're currently in custody in Delius?

Reply:
As it is, this will throw an error if someone is matched while in the community.  Which could happen quite often?

Reply:
Happy to have a quick chat about it, I may be missing something - it feels like we could continue with a null/unknown prison, then rely on the validation method to log an UnexpectedInstitution event as normal.  We wouldn't get alerted for the exception, but not the end of the world.